packages/cache/src/cache.ts

```diff
@@ -0,0 +1,169 @@
+ */
+export async function restoreCache(
+  path: string,
+  primaryKey: string,
+  restoreKeys?: string[]
+): Promise<string | undefined> {
+  try {
+    if (!path || path.length === 0) {
+      throw new Error('Input required and not supplied: path')
+    }
+
+    restoreKeys = restoreKeys || []
+    const keys = [primaryKey, ...restoreKeys]
+
+    core.debug('Resolved Keys:')
+    core.debug(JSON.stringify(keys))
+
+    if (keys.length > 10) {
+      core.setFailed(
+        `Key Validation Error: Keys are limited to a maximum of 10.`
+      )
+      return undefined
+    }
+    for (const key of keys) {
+      if (key.length > 512) {
+        core.setFailed(
+          `Key Validation Error: ${key} cannot be larger than 512 characters.`
+        )
+        return undefined
+      }
+      const regex = /^[^,]*$/
+      if (!regex.test(key)) {
+        core.setFailed(`Key Validation Error: ${key} cannot contain commas.`)
+        return undefined
+      }
+    }
+
+    const compressionMethod = await utils.getCompressionMethod()
+
+    try {
+      // path are needed to compute version
+      const cacheEntry = await cacheHttpClient.getCacheEntry(keys, path, {
+        compressionMethod
+      })
+      if (!cacheEntry?.archiveLocation) {
+        core.info(`Cache not found for input keys: ${keys.join(', ')}`)
+        return undefined
+      }
+
+      const archivePath = pathUtils.join(
+        await utils.createTempDirectory(),
+        utils.getCacheFileName(compressionMethod)
+      )
+      core.debug(`Archive Path: ${archivePath}`)
+
+      try {
+        // Download the cache from the cache entry
+        await cacheHttpClient.downloadCache(
+          cacheEntry.archiveLocation,
+          archivePath
+        )
+
+        const archiveFileSize = utils.getArchiveFileSize(archivePath)
+        core.info(
+          `Cache Size: ~${Math.round(
+            archiveFileSize / (1024 * 1024)
+          )} MB (${archiveFileSize} B)`
+        )
+
+        await extractTar(archivePath, compressionMethod)
+      } finally {
+        // Try to delete the archive to save space
+        try {
+          await utils.unlinkFile(archivePath)
+        } catch (error) {
+          core.debug(`Failed to delete archive: ${error}`)
+        }
+      }
+
+      core.info(`Cache restored from key: ${cacheEntry && cacheEntry.cacheKey}`)
+
+      return cacheEntry.cacheKey
+    } catch (error) {
+      utils.logWarning(error.message)
+      return undefined
+    }
+  } catch (error) {
+    core.setFailed(error.message)
+    return undefined
+  }
+}
+
+/**
+ * Saves a file with the specified key
+ *
+ * @param path a string representing files to be cached
+ * @param key an explicit key for restoring the cache
+ * @returns number returns cacheId if the cache was saved successfully, otherwise return -1
+ */
+export async function saveCache(path: string, key: string): Promise<number> {
```

Comment:
We should support multiple paths for this input, we don't have to rely on splitting `path` by a newline in the package itself.
```suggestion
export async function saveCache(paths: string[], key: string): Promise<number> {
```